### PR TITLE
[f] Application generator

### LINKED ISSUE #157: Application generator
**Taker**: @hieuk09 

## What

Introduce a new command `lotus generate`.

## Why

Lotus generator (`lotus new`) generates only one application `apps/web`. If developers want to add other apps, this will be the official way to do that.

### CLI

```shell
% lotus generate app admin
```

  * `app` is the subject of what we want to generate
  * `admin` is the name of the application under `apps/`
  * Optional `--application-base-url`. Defaults to the app name.
  * Optional `--test`. Defaults to the contents of `.lotusrc` or to `minitest`, if that file is missing.

### Output

  * It will generate a Lotus application with the given name: `apps/admin`.
  * Amend `config/environment.rb`

### Files

```ruby
# config/environment.rb
require 'rubygems'
require 'bundler/setup'
require 'lotus/setup'
require_relative '../lib/bookshelf'
require_relative '../apps/web/application'
require_relative '../apps/admin/application' # NEW LINE

Lotus::Container.configure do
  #
  # !!! ATTENTION !!! New applications should be put at the top of this block
  #
  mount Admin::Application, at: '/admin' # NEW LINE
  mount Web::Application, at: '/'
end
```

### PR DESCRIPTION
**Pull request by hieuk09**
[Close #157]

### CODE REVIEW

File: lib/lotus/commands/generate.rb [hunk 1]
```diff
@@ -24,16 +24,21 @@ class Error < ::StandardError
...
         @cli      = cli
-        @options  = env.to_options.merge(cli.options)
-
-        @app_name = app_name
-        @app      = Utils::String.new(@app_name).classify
-
         @name     = name
-        @type     = type
 
+        if type == 'app'
```
Comment (nguyenngoc2505) [inline lib/lotus/commands/generate.rb]: Please remove this

File: test/integration/cli/generate_test.rb [hunk 2]
```diff
@@ -50,20 +55,12 @@ def chdir_to_root
       generate_application
       generate_action
       generate_action_without_view
-      generate_model
```
Comment (nguyenngoc2505) [inline test/integration/cli/generate_test.rb]: why do you remove this test function?

File: lib/lotus/commands/generate.rb (hunk 1 above)
Comment (hieuk09) [inline lib/lotus/commands/generate.rb]: Fixed, thanks.

File: lib/lotus/commands/generate.rb [hunk 3]
```diff
@@ -72,6 +71,21 @@ def generator
...
+        {
+          application: app_name,
+          application_base_url: "/#{app_name}"
+        }
+      end
+
+      # @since x.x.x
+      # @api private
+      def sanitize_type(type)
+        type == 'app' ? 'slice' : type
```
Comment (AlfonsoUceda) [inline lib/lotus/commands/generate.rb]: can you extract these strings to constants?
```ruby
APP = 'app'.freeze
SLICE_TYPE = 'slice'.freeze
```
something like that

**Comment by AlfonsoUceda** [comment]
:+1:

File: lib/lotus/commands/generate.rb (hunk 3 above)
Comment (hieuk09) [inline lib/lotus/commands/generate.rb]: Fixed, thanks.

File: lib/lotus/commands/generate.rb [hunk 4]
```diff
@@ -10,6 +10,8 @@ class Generate
       # @since 0.3.0
       # @api private
       GENERATORS_NAMESPACE = "Lotus::Generators::%s".freeze
+      APP = 'app'.freeze
+      SLICE_TYPE = 'slice'.freeze
```
Comment (joneslee85) [inline lib/lotus/commands/generate.rb]: @jodosha are you okay with this term?

**Comment by joneslee85** [comment]
:+1:

File: lib/lotus/commands/generate.rb (hunk 4 above)
Comment (jodosha) [inline lib/lotus/commands/generate.rb]: @joneslee85 yup :+1:

**Comment by jodosha** [comment]
@hieuk09 Thanks for this PR, we're on the right way.

We should remove `test` CLI option. This decision should be taken by looking at `.lotusrc`. Please note that `Lotus::Environment.new.to_options` will lookup that file for us.

Also, can you please change the name of the generated application from `api` to `admin`? The reason is that in the near future we will be able to do `lotus generate api my_api`, so the name of the fixtures will not be confusing.

Thank you! /cc @joneslee85

**Comment by hieuk09** [comment]
Thanks for your comments, @jodosha. I'll update the PR accordingly.

**Comment by hieuk09** [comment]
@jodosha: I fixed the PR as your comments.

**Comment by joneslee85** [comment]
:+1: nice one

**Comment by jodosha** [comment]
@hieuk09 Thank you very much. :confetti_ball: Merging it! :+1:

**Pull request merged**